zrml/rikiddo/src/types.rs

```diff
@@ -29,21 +33,67 @@ pub enum Timespan {
             Timespan::Seconds(d) => d,
-            Timespan::Minutes(d) => d * 60,
-            Timespan::Hours(d) => d * 60 * 60,
-            Timespan::Days(d) => u32::from(d) * 60 * 60 * 24,
-            Timespan::Weeks(d) => u32::from(d) * 60 * 60 * 24 * 7,
+            Timespan::Minutes(d) => d.saturating_mul(60),
+            Timespan::Hours(d) => d.saturating_mul(60).saturating_mul(60),
+            Timespan::Days(d) => {
+                u32::from(d).saturating_mul(60).saturating_mul(60).saturating_mul(24)
+            }
+            Timespan::Weeks(d) => u32::from(d)
+                .saturating_mul(60)
+                .saturating_mul(60)
+                .saturating_mul(24)
+                .saturating_mul(7),
         }
     }
 }
 
-/*
-#[derive(Clone, Debug, Decode, Encode, Eq, PartialEq)]
-pub struct RikiddoSigmoidMV<FI: Fixed, FE: Sigmoid<FI>, MA: MarketAverage<FI>> {
-    pub fees: FE,
-    pub ma_short: MA,
-    pub ma_long: MA,
-    pub _marker: PhantomData<FI>,
+/// Returns integer part of FROM, if the msb is not set and and num does it into FROM
+fn convert_common<FROM: Fixed, TO: Fixed>(num: FROM) -> Result<TO, &'static str> {
+    // Check if number is negatie
+    if num < FROM::from_num(0u8) {
+        return Err("Cannot convert negative signed number into unsigned number");
+    }
+
+    // PartialOrd is bugged, therefore the workaround
+    // https://github.com/encointer/substrate-fixed/issues/9
+    let num_u128: u128 = num.int().to_num();
+    if TO::max_value().int().to_num::<u128>() < num_u128 {
+        return Err("Fixed point conversion failed: FROM type does not fit in TO type");
+    }
+
+    Ok(num_u128.to_fixed())
+}
+
+/// Converts an unsigned fixed point number into a signed fixed point number (fallible)
+pub fn convert_to_signed<FROM: FixedUnsigned, TO: FixedSigned + LossyFrom<FixedI128<U127>>>(
+    num: FROM,
+) -> Result<TO, &'static str> {
+    let integer_part: TO = convert_common(num)?;
+    let fractional_part: FixedI128<U127> = num.frac().to_fixed();
+
+    if let Some(res) = integer_part.checked_add(fractional_part.lossy_into()) {
+        Ok(res)
+    } else {
+        // This error should be impossible to reach.
+        Err("Something went wrong during FixedUnsigned to FixedSigned type conversion")
+    }
+}
+
+/// Converts a signed fixed point number into an unsigned fixed point number (fallible)
+pub fn convert_to_unsigned<FROM: FixedSigned, TO: FixedUnsigned + LossyFrom<FixedU128<U128>>>(
+    num: FROM,
+) -> Result<TO, &'static str> {
+    // We can safely cast because until here we know that the msb is not set.
+    let integer_part: TO = convert_common(num)?;
+    let fractional_part: FixedU128<U128> = num.frac().to_fixed();
+
+    if let Some(res) = integer_part.checked_add(fractional_part.lossy_into()) {
+        Ok(res)
+    } else {
+        // This error should be impossible to reach.
+        Err("Something went wrong during FixedSigned to FixedUnsigned type conversion")
+    }
```

Comment:
Those can be potentially implemented in a more stylish fashion for all adequate Fixed types by using a blanked implementation. I guess it is fine like this, and the blanket implementation could be a low priority enhancement.

Reply:
agreed